src/tcp/ClickHouseSock.jl

```diff
@@ -0,0 +1,102 @@
+Base.@kwdef struct CHSettings
+    host::String = "localhost"
+    port::Int = 9000
+    database::String = ""
+    username::String = "default"
+    password::String = ""
+    connection_timeout::Int = DBMS_DEFAULT_CONNECT_TIMEOUT
+    max_insert_block_size::Int = DBMS_DEFAULT_MAX_INSERT_BLOCK
+    send_buffer_size::Int = DBMS_DEFAULT_BUFFER_SIZE
+end
+
+
+mutable struct ClickHouseSock
+    io ::Union{IO, Nothing}
+    settings ::CHSettings
+    cond ::ReentrantLock
+    busy ::Bool
+    server_name ::String
+    server_rev ::Int
+    server_timezone ::Union{String, Nothing}
+
+    function ClickHouseSock(io, settings = CHSettings())
+        return new(
+            io,
+            settings,
+            ReentrantLock(),
+            false,
+            "", 0, nothing
+        )
+    end
+end
+
+function guarded(f, sock::ClickHouseSock)
+    lock(sock.cond)
+    try
+        res = f(sock)
+        unlock(sock.cond)
+        return res
+    catch e
+        unlock(sock.cond)
+        rethrow(e)
+    end
+end
+
+macro guarded(sock, expr)
+    quote
+        lock($(esc(sock)).cond)
+        local res = try
+            $(esc(expr))
+        catch e
+            unlock($(esc(sock)).cond)
+            rethrow(e)
+        end
+        unlock($(esc(sock)).cond)
+        res
+    end
+end
+
+is_connected(sock::ClickHouseSock) = @guarded sock !isnothing(sock.io)
+is_busy(sock::ClickHouseSock) = @guarded sock sock.busy
+
+set_busy!(sock::ClickHouseSock, value::Bool) =
+                        @guarded sock sock.busy = value
+
+function using_socket(f, sock::ClickHouseSock)
```

Comment:
Same as with `guarded` -- why two versions? Other than with `guarded`, this one has a few references for both versions.

Reply:
done at df9f9ed